### PR TITLE
Sroute

### PR DESCRIPTION
**Pull request by Wen-Tian-Pineapple**
update for sroute command

### CODE REVIEW

File: src/odb/include/odb/db.h [hunk 1]
```diff
@@ -38,6 +38,7 @@
 #include <string>
 #include <vector>
 
+#include "ISdb.h"
```
Comment (maliberty) [inline src/odb/include/odb/db.h:41]: Why is this needed?

File: src/odb/include/odb/db.h [hunk 2]
```diff
@@ -9299,6 +9298,14 @@ class dbGCellGrid : public dbObject
   // User Code End dbGCellGrid
 };
 
+class dbPGpin : public dbObject
+{
+  public:
+    void create_PGpin (dbBlock* block, dbTech* tech, const char* source_net_name, int num_connection_points, std::string position);
```
Comment (maliberty) [inline src/odb/include/odb/db.h:9304]: This should just be called create as with the other classes.

**Comment by Wen-Tian-Pineapple** [comment]
> 
yes, I just did the modification that the bot suggested

File: src/odb/src/db/dbPGpin.cpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+void dbPGpin::create_PGpin (dbBlock* block, dbTech* tech, const char * source_net_name, int num_connection_points, const std::string& position) {
+  /*
+  _dbPGpin* obj = (_dbPGpin*) this;
+  dbDatabase* db = (dbDatabase*)obj->getDatabase();
+  
+  db->getTech();
+  dbBlock* block = db->getChip()->getBlock();
+  dbTech* tech = db->getTech();
+  */
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:91]: remove

File: src/odb/src/db/dbPGpin.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+  dbNet* create_pin_net = block->findNet(source_net_name);
+  dbTechLayer* PGpin_metal_layer = tech->findLayer("met1");
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:94]: hard coded name?  I guess you want to take a dbLayer not a dbTech as the arg

File: src/odb/include/odb/db.h [hunk 5]
```diff
@@ -9299,6 +9298,14 @@ class dbGCellGrid : public dbObject
   // User Code End dbGCellGrid
 };
 
+class dbPGpin : public dbObject
```
Comment (maliberty) [inline src/odb/include/odb/db.h:9301]: Please give a class level comment explaining the object.

File: src/odb/src/db/dbPGpin.cpp [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+void dbPGpin::create_custom_connections (dbBlock* block, const char* nett, const char* instt, const char* itermm) {
+  dbNet* net = block->findNet(nett);
+
+  dbInst* inst = block->findInst(instt);
+  dbITerm* iterm = inst->findITerm(itermm);
+  iterm->connect(net);
+
+}
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:220]: This has nothing to do with the dbPGpin object.  Why is it a method?  Why all the duplicate letters in nett, instt, termm?

File: src/odb/src/db/dbPGpin.cpp [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+  // User Code End Constructor
+}
+
+_dbPGpin::~_dbPGpin()
+{
+  // User Code Begin Destructor
+  // User Code End Destructor
+}
+
+int r2grid (int number, dbTech *db_tech) {
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:75]: add static for file scope
Improve the name, eg snapToGrid

File: src/odb/src/db/dbPGpin.cpp [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+  /*
+  _dbPGpin* obj = (_dbPGpin*) this;
+  dbDatabase* db = (dbDatabase*)obj->getDatabase();
+  
+  db->getTech();
+  dbBlock* block = db->getChip()->getBlock();
+  dbTech* tech = db->getTech();
+  */
+
+  dbNet* create_pin_net = block->findNet(source_net_name);
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:93]: Why doesn't this take a dbNet instead of its name?

File: src/odb/src/db/dbPGpin.cpp [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+  dbTechLayer* PGpin_metal_layer = tech->findLayer("met1"); 
+  dbSet<dbSWire> swires = create_pin_net->getSWires();
+  dbSBox* pdn_wire = new dbSBox();
+
+  int num = 0;
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:98]: This appears to be a boolean - perhaps bool first = true?

**Comment by Wen-Tian-Pineapple** [comment]
I just reverse the change previous the bot suggested

File: src/odb/src/db/dbPGpin.cpp [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+  int direction_0;
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:99]: Why _0?  There is no _1...

File: src/odb/src/db/dbPGpin.cpp [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+  //find highest metal layer
+  for(auto swire : swires){
+    for (auto wire : swire->getWires()) {
+      if (num == 0){
+        num++;
+        PGpin_metal_layer = wire->getTechLayer();
+        pdn_wire = wire;
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:106]: Why did you create a dbSBox above?  It seems you just throw it away here.

File: src/odb/src/db/dbPGpin.cpp [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+            }
+          }
+        }
+      }
+    }
+  }
+  
+  
+  std::string r_pin = "pg_" + std::string(source_net_name);
+  dbNet* r_net = dbNet::create(block, r_pin.c_str(), true);
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:132]: Why are you skipping the duplicate check?

File: src/odb/src/db/dbPGpin.cpp [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+        xMax[n] = xMin[n] + dx;
+
+      } else if (position == "default") {
+        int dx = r2grid((pdn_wire->xMax() - pdn_wire->xMin())/num_connection_points, tech);
+        xMin[n] = pdn_wire->xMin() + n*dx;
+        xMax[n] = xMin[n] + dx;
+
+      } else {
+        std::cout << "WARNING.";
+        break;
```
Comment (maliberty) [inline src/odb/src/db/dbPGpin.cpp:173]: Use the logger for messages (other places as well).  This seems like an error

File: src/odb/src/swig/common/swig_common.cpp [hunk 14]
```diff
@@ -26,7 +27,7 @@
  */
 
 #include "swig_common.h"
-
+#include "../../db/dbPGpin.h"
```
Comment (maliberty) [inline src/odb/src/swig/common/swig_common.cpp:30]: You should not be swigging the private header.

File: src/odb/src/swig/common/swig_common.cpp [hunk 15]
```diff
@@ -298,6 +299,18 @@ void createSBoxes(odb::dbSWire* swire,
     odb::dbSBox::create(swire, via, point.getX(), point.getY(), type);
 }
 
+void createPGpins(odb::dbBlock* block, odb::dbTech* tech, const char* source_net_name, int num_connection_points, std::string position)
+{
+  odb::dbPGpin PGpin;
+  PGpin.create_PGpin(block, tech, source_net_name, num_connection_points, position);
+}
+
+void p2proute(odb::dbBlock* block, const char* nett, const char* instt, const char* itermm)
```
Comment (maliberty) [inline src/odb/src/swig/common/swig_common.cpp:308]: This doesn't do any routing.  It is more about building a connection.

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/odb/src/db/dbPGpin.h [hunk 16]
```diff
@@ -0,0 +1,70 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (c) 2020, The Regents of the University of California
```
Comment (rovinski) [inline src/odb/src/db/dbPGpin.h:4]: 2023

File: src/odb/src/db/dbPGpin.cpp (hunk 4 above)
Comment (Wen-Tian-Pineapple) [reply src/odb/src/db/dbPGpin.cpp:94]: actually PGpin_metal_layer doesn't need to get instantiated here so I deleted it. It's for search the highest layer later

File: src/odb/src/swig/common/swig_common.cpp (hunk 15 above)
Comment (Wen-Tian-Pineapple) [reply src/odb/src/swig/common/swig_common.cpp:308]: name changed

File: src/odb/src/db/dbPGpin.h (hunk 16 above)
Comment (QuantamHD) [reply src/odb/src/db/dbPGpin.h:4]: @Wen-Tian-Pineapple you should include your employer ("The Regents of the University of Michigan") in the copyright header not the University of California.

File: src/odb/src/db/dbPGpin.cpp (hunk 10 above)
Comment (Wen-Tian-Pineapple) [reply src/odb/src/db/dbPGpin.cpp:99]: renamed

File: src/odb/src/db/dbPGpin.cpp (hunk 9 above)
Comment (Wen-Tian-Pineapple) [reply src/odb/src/db/dbPGpin.cpp:98]: changed

File: src/odb/src/db/dbPGpin.cpp (hunk 8 above)
Comment (Wen-Tian-Pineapple) [reply src/odb/src/db/dbPGpin.cpp:93]: because we need to use the getSWires function of it

File: src/odb/src/db/dbPGpin.cpp [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+
+      } else if (position == Position::MIDDLE) {
+        int dy = snapToGrid((pdn_wire->getDY())/num_connection_points/2, tech);
+        yMin[n] = pdn_wire->yMin() + (pdn_wire->getDY())/4 + n*dy;
+        yMax[n] = yMin[n] + dy;
+
+      } else if (position == Position::DEFAULT) {
+        int dy = snapToGrid((pdn_wire->getDY())/num_connection_points, tech);
+        yMin[n] = pdn_wire->yMin() + n*dy;
+        yMax[n] = yMin[n] + dy;
```
Comment (rovinski) [inline src/odb/src/db/dbPGpin.cpp:191]: `switch` statements are preferred over `if else` for enums

File: src/odb/src/db/dbPGpin.cpp [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+      } else {
+        logger->error(utl::ODB,
+                   1103,"wrong position.");
+        break;
```
Comment (rovinski) [inline src/odb/src/db/dbPGpin.cpp:196]: `break` is unnecessary because `error()` will throw an exception.

File: src/odb/src/db/dbUtil.cpp [hunk 19]
```diff
@@ -1626,4 +1626,158 @@ bool dbCreateNetUtil::createSingleVia(dbNet* net,
   return true;
 }
 
+static int snapToGrid (int number, dbTech *db_tech) {
+    // This will round "number" to a value that respects the manufacturing grid
+    // For example, if the manufacturing grid is 5 microns and number = 20333,
+    // the procedure will output 20335.
+    auto grid = db_tech->getManufacturingGrid();
+    return round(number / double(grid) * grid);
```
Comment (maliberty) [inline src/odb/src/db/dbUtil.cpp:1634]: This is incorrect, use:

return round(number / double(grid)) * grid;

File: src/odb/src/db/dbUtil.cpp [hunk 20]
```diff
@@ -1626,4 +1626,158 @@ bool dbCreateNetUtil::createSingleVia(dbNet* net,
...
+}
+
+void dbCreateNetUtil::createPGpin (dbBlock* block, dbTech* tech, const char * source_net_name, int num_connection_points, Position position) {
+
+  utl::Logger* logger = new utl::Logger();
```
Comment (maliberty) [inline src/odb/src/db/dbUtil.cpp:1639]: Don't create a new Logger but use the one from the appplication.

File: src/odb/src/db/dbUtil.cpp [hunk 21]
```diff
@@ -1626,4 +1626,158 @@ bool dbCreateNetUtil::createSingleVia(dbNet* net,
...
 
+static int snapToGrid (int number, dbTech *db_tech) {
+    // This will round "number" to a value that respects the manufacturing grid
+    // For example, if the manufacturing grid is 5 microns and number = 20333,
+    // the procedure will output 20335.
+    auto grid = db_tech->getManufacturingGrid();
+    return round(number / double(grid) * grid);
+}
+
+void dbCreateNetUtil::createPGpin (dbBlock* block, dbTech* tech, const char * source_net_name, int num_connection_points, Position position) {
```
Comment (maliberty) [inline src/odb/src/db/dbUtil.cpp:1637]: clang-format

File: src/odb/src/db/dbUtil.cpp [hunk 22]
```diff
@@ -1626,4 +1626,158 @@ bool dbCreateNetUtil::createSingleVia(dbNet* net,
...
+  //find highest metal layer
+  for(auto swire : swires){
+    for (auto wire : swire->getWires()) {
+      if (first){
+        first = false;
+        PGpin_metal_layer = wire->getTechLayer();
+        pdn_wire = wire;
+      }
+      else {
+        if (wire->getTechLayer() > PGpin_metal_layer ) {
```
Comment (maliberty) [inline src/odb/src/db/dbUtil.cpp:1656]: No need to nest here, use else if

File: src/odb/src/db/dbUtil.cpp [hunk 23]
```diff
@@ -1626,4 +1626,158 @@ bool dbCreateNetUtil::createSingleVia(dbNet* net,
...
+  for (int n = 0; n < num_connection_points; n++) {
+    
+    dbBTerm * r_bterm = dbBTerm::create(r_net, (r_pin+ "_" + std::to_string(n)).c_str());
+    dbBPin * r_bpin = dbBPin::create(r_bterm);
+    r_bpin->setPlacementStatus("FIRM");
+
+    direction = pdn_wire->getDir();
+    if ((direction == 1) | (direction == -1)) {
+      // for horizontal
+      switch (position) {
```
Comment (maliberty) [inline src/odb/src/db/dbUtil.cpp:1696]: missing break statements throughout

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/odb/src/swig/common/dbhelpers.i [hunk 24]
```diff
@@ -53,5 +53,11 @@ void createSBoxes(odb::dbSWire* swire,
                   std::vector<odb::Point> points,
                   odb::dbWireShapeType type);
 
+void createPGpins(odb::dbBlock* block, odb::dbTech* tech, const char* source_net_name, 
+                  int num_connection_points, 
+                  int position);
+
+void createConnection(odb::dbBlock* block, const char* net, const char* inst, const char* iterm);
```
[the remark this answers is not in the document]
Comment (Wen-Tian-Pineapple) [reply src/odb/src/swig/common/dbhelpers.i:60]: yes it it

File: src/odb/src/db/dbUtil.cpp (hunk 19 above)
Comment (Wen-Tian-Pineapple) [reply src/odb/src/db/dbUtil.cpp:1634]: modified

File: src/odb/src/swig/common/dbhelpers.i (hunk 24 above)
[the remark this answers is not in the document]
Comment (maliberty) [reply src/odb/src/swig/common/dbhelpers.i]: If it is required there should be a test.

**Comment by maliberty** [comment]
replaced by https://github.com/The-OpenROAD-Project/OpenROAD/pull/3551
